src/lib/support/jsontlv/TlvToJson.cpp

```diff
@@ -0,0 +1,371 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/support/CHIPMemString.h"
+#include "lib/support/ScopedBuffer.h"
+#include <json/json.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/jsontlv/ElementTypes.h>
+#include <lib/support/jsontlv/TlvToJson.h>
+
+namespace chip {
+
+namespace {
+
+/*
+ * Encapsulates the different types of keys permissible.
+ *
+ * Root Key = Key with a name of 'value'. This is the top-most key in a given JSON object generated from TLV.
+ * Struct Field = Key containing the 32-bit field ID of an item in a struct.
+ * Array Item = Key containing the 16-bit list index of an item in a list.
+ *
+ * In the latter two modes, the actual field ID/list index is encapsulated within the 'key' member.
+ *
+ */
+struct KeyContext
+{
+    enum ContainerType
+    {
+        kStruct,
+        kArray
+    };
+
+    KeyContext() = default;
+
+    KeyContext(chip::FieldId fieldId)
+    {
+        containerType = kStruct;
+        tag           = fieldId;
+    }
+
+    KeyContext(chip::ListIndex listIndex)
+    {
+        containerType = kArray;
+        tag           = listIndex;
+    }
+
+    void SetElementType(const char * elemType) { elementType = elemType; }
+
+    void SetArrayElementType(const char * subType) { subElementType = subType; }
+
+    void GenerateJsonKeyStr(char * buf) const
+    {
+        int len = 0;
+        len     = sprintf(buf, "%u:", tag);
+        len += sprintf(&buf[len], "%s", elementType);
+        if (strcmp(elementType, kElementTypeArray) == 0)
+        {
+            sprintf(&buf[len], "-%s", subElementType);
+        }
+    }
+
+    ContainerType containerType = kStruct;
+    unsigned int tag            = 0;
+    const char * elementType    = nullptr;
+    const char * subElementType = nullptr;
+};
+
+/*
+ * For now, let's put a bound of the maximum length of a byte/char string to be the size of an IPv6
+ * MTU. While this is smaller than that of the limit defined in the data model specification,
+ * strings by virtue of not being chunked are intrinsically limited in size to the size of the encompassing packet.
+ */
+static constexpr uint16_t kMaxStringLen = 1280;
+
+/*
+ * This templated function inserts a key/value pair into the Json value object.
+ * The value is templated to be of type T and accepts any of the following primitive
+ * types:
+ *      bool, uint*_t, int*_t, char *, float, double.
```

Comment:
```suggestion
 *      bool, uint*_t, int*_t, char *, float, double, std::string, Json::Value
```
though I question how "primitive" some of those are.

Reply:
updated